### PR TITLE
Added support for LIKE queries to predicate parser

### PR DESCRIPTION
**Pull request by astigsen**
This was tested with the js binding as well.

@bdash @alazier

### CODE REVIEW

File: tests/parser.cpp [hunk 1]
```diff
@@ -80,6 +80,9 @@ static std::vector<std::string> valid_queries = {
     "contains contains 'contains'",
     "beginswith beginswith 'beginswith'",
     "endswith endswith 'endswith'",
+    "a like '*'",
```
Comment (bdash) [inline tests/parser.cpp]: We should probably test that column names of `like` still work, similar to what we do above for `contains` / `beginswith` / `endswith`.

File: tests/query.json [hunk 2]
```diff
@@ -201,6 +203,7 @@
         ["QueryCount", 5, "StringObject", "stringCol BEGINSWITH[c] 'A'"],
         ["QueryCount", 4, "StringObject", "stringCol ENDSWITH[c] 'c'"],
         ["QueryCount", 2, "StringObject", "stringCol CONTAINS[c] 'B'"]
+        ["QueryCount", 2, "StringObject", "stringCol LIKE[c] '*B?'"]
```
Comment (bdash) [inline tests/query.json]: It'd be good to test handling of invalid predicates like `stringCol LIKE 1`, and substituted patterns like `stringCol LIKE $0`

**Comment by austinzheng** [comment]
This seems useful. Should we revive it?

**Comment by bdash** [comment]
At some point, yes. Someone just needs to spend the time expanding the test coverage of this new functionality.

**Comment by austinzheng** [comment]
If tests are the only missing thing, I'll see what I can do later this week. Seems pretty straightforward.

**Comment by austinzheng** [comment]
I'm writing tests; when I get home and confirm they work I'll update this PR.

**Comment by austinzheng** [comment]
I'm using CI to check my work, so I'll wait until it succeeds to merge.

File: tests/query.json [hunk 3]
```diff
@@ -200,7 +209,13 @@
         ["QueryCount", 3, "StringObject", "stringCol ==[c] 'a'"],
         ["QueryCount", 5, "StringObject", "stringCol BEGINSWITH[c] 'A'"],
         ["QueryCount", 4, "StringObject", "stringCol ENDSWITH[c] 'c'"],
-        ["QueryCount", 2, "StringObject", "stringCol CONTAINS[c] 'B'"]
+        ["QueryCount", 2, "StringObject", "stringCol CONTAINS[c] 'B'"],
+        ["QueryCount", 2, "StringObject", "stringCol LIKE[c] '*B?'"],
+        ["QueryCount", 0, "StringObject", "stringCol LIKE 1"],
+        ["QueryCount", 0, "StringObject", "stringCol LIKE LIKE"],
+        ["QueryCount", 0, "StringObject", "stringCol LIKE[A] '*BC'"],
```
Comment (bdash) [inline tests/query.json]: What does `[A]` do here? Does this query execute successfully?

Comment (austinzheng) [inline tests/query.json]: If it doesn't, I'll remove it and add it to the list of malformed queries in `parser.cpp`. It seems that if `"stringCol LIKE 1"` parses then something like this should parse too. I need to double-check.

Comment (austinzheng) [inline tests/query.json]: (Sorry, I didn't answer your question. It's an invalid flag and the query should return 0 results.)

Comment (bdash) [inline tests/query.json]: `[c]` / `[C]` is the only modifier that's accepted based on https://github.com/realm/realm-object-store/blob/86138ce4e55ca04e975c1165d409f02dcb729b6b/src/parser/parser.cpp#L73

I'd expect `stringCol LIKE 1` to _parse_ successfully but be rejected when we attempt to build a query from it due to the type mismatch.

Comment (bdash) [inline tests/query.json]: Hmm, I'm surprised that we don't distinguish between a query that matched 0 objects and an error (I'm assuming an exception is being thrown).

Comment (austinzheng) [inline tests/query.json]: I moved the bad test cases into `parser.cpp`'s bad test case list, and the tests passed, so you were right.

I'm actually not sure how `query.json` fits into the tests. It isn't referenced anywhere else within the object store project, and breaking the JSON doesn't affect the test outcomes.

Comment (bdash) [inline tests/query.json]: It looks like query.json is used by the JS binding. Not entirely sure why it lives in the object store repo.

**Comment by austinzheng** [comment]
I'll figure out how to run the `json` test file from the JS binding tonight or Monday. I want to ensure everything is passing before merging anything else.

**Comment by bdash** [comment]
I suspect @ironage took care of this after moving the query parser to core.

**Pull request closed without merging**